core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateAction.java

```diff
@@ -85,21 +88,24 @@
     private final DiscoverySettings discoverySettings;
     private final ClusterName clusterName;
     private final PendingClusterStatesQueue pendingStatesQueue;
+    private final CustomPrototypeRegistry registry;
 
     public PublishClusterStateAction(
             Settings settings,
             TransportService transportService,
             Supplier<ClusterState> clusterStateSupplier,
             NewPendingClusterStateListener listener,
             DiscoverySettings discoverySettings,
-            ClusterName clusterName) {
+            ClusterName clusterName,
+            CustomPrototypeRegistry registry) {
```

Comment:
I wonder if we should pass these registries all over the place? maybe it's better to give a cluster diff and cluster state deserializers ? It doesn't feel like PublishClusterStateAction should care about the registry.